site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+
+	// Rules are a list of GRPC matchers, filters and actions.
+        // 
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:default={{matches: {{method: {type: "Exact"}}}}}
+	Rules []GRPCRouteRule `json:"rules,omitempty"`
+}
+
+// GRPCRouteRule defines semantics for matching an gRPC request based on
+// conditions (matches), processing it (filters), and forwarding the request to
+// an API object (backendRefs).
+type GRPCRouteRule struct {
+	// Matches define conditions used for matching the rule against incoming
+	// gRPC requests. Each match is independent, i.e. this rule will be matched
+	// if **any** one of the matches is satisfied.
+	//
+	// For example, take the following matches configuration:
+	//
+	// ```
+	// matches:
+	// - method:
+	//     method: foo.bar
+	//   headers:
+	//     values:
+	//       version: 2
+	// - method:
+	//     service: foo.bar.v2
+	// ```
+	//
+	// For a request to match against this rule, a request should satisfy
+	// EITHER of the two conditions:
+	//
+	// - service of foo.bar AND contains the header `version: 2`
+	// - service of foo.bar.v2
+	//
+	// See the documentation for GRPCRouteMatch on how to specify multiple
+	// match conditions that should be ANDed together.
+	//
+	// If no matches are specified, the matches every gRPC request.
+	//
+	// Proxy or Load Balancer routing configuration generated from GRPCRoutes or HTTPRoutes
+	// MUST prioritize rules based on the following criteria, continuing on
+	// ties. When comparing GRPCRoutes and HTTPRoutes, GRPCRoute matches are first converted
+	// to HTTPRoute matches according to the following scheme:
+	//
+	// * HeaderMatches are translated exactly, since the field are
+	//    one-to-one
+	// * If `MethodMatch.type` is `Exact` and neither the
+	//    `MethodMatch.service` nor `MethodMatch.method` fields are present,
+	//    translate to a `PathMatch` of `{"type": "PathPrefix", value: "/"}`
+	// * If `MethodMatch.type` is `Exact` and both the `MethodMatch.service`
+	//    and `MethodMatch.method` fields are present, translate to a `PathMatch`
+	//    of `{"type": "Exact", value: "/" + service + "/" + method}`
+	// * If `MethodMatch.type` is `Exact` and the `MethodMatch.service`
+	//    field is present but not the `MethodMatch.method` field, translate to a
+	//    `PathMatch` of `{"type": "PathPrefix", value: "/" + service}`
+	// - Otherwise, translate the service and the method individually into
+	//    regular expressions (with an absent field interpreted as `.*`) and
+	//    translate to a `PathMatch` of
+	//    `{"type": "RegularExpression", value: "/": serviceRegex + "/" + methodRegex}`
+	//
+	// Once the translation is applied, precedence must be given to the Rule with
+	// the largest number of:
+	//
+	// * Characters in a matching non-wildcard hostname.
+	// * Characters in a matching hostname.
+        // * Characters in a matching path.
+	// * Header matches.
+        // * Query param matches.
```

Comment:
No query-parameter matching is supported yet.
```suggestion
```

Reply:
Hm. This was pulled from the HTTPRoute docstrings. Removed.